sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsReflectionSetter.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.options;
+
+import java.beans.IntrospectionException;
+import java.beans.Introspector;
+import java.beans.PropertyDescriptor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import org.apache.beam.sdk.util.StringUtils;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableListMultimap;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.Iterables;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ListMultimap;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.Sets;
+
+/** This is a utility class to set and remove options individually. */
+public class PipelineOptionsReflectionSetter {
+  private static final boolean STRICT_PARSING = true;
+
+  @SuppressWarnings("unchecked")
+  public static Class<? extends PipelineOptions> getPipelineOptionsInterface(
+      PipelineOptions options) {
+    if (options.getClass().getInterfaces().length != 1) {
```

Comment:
PipelineOptions interface validation is fairly strict on making sure that methods have the same default/parameter type/...
You should really support multiple interfaces since its quite common in our codebase.

Reply:
Do you mean a pipeline option interface that extends multiple interfaces or the object is implementing multiple interfaces? I couldn't find any example that the pipeline option object is implementing two interfaces. At least if the user uses PipelineOptionsFactory.as() or PipelineOptionsFactory.create() it will return a proxy object that only implements the given interface. (The interface though can be subclass of multiple interfaces itself, and that would be OK.)

Reply:
Now that I know your intent, you should instead check to see if the class is an interface using:
https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#isInterface--

From `Class#getInterfaces`:
If this object represents an interface, the array contains objects representing all interfaces extended by the interface. 

And hence would fail on things like DataflowPipelineOptions which extends multiple interfaces.

Reply:
If I call options.getClass() it will not return DataflowPipelineOptions, it will return a proxy class. For instance BigQueryOptions also extends multiple interfaces (Similar to DataflowPipelineOptions); however, if I run the following code:

    BigQueryOptions options = PipelineOptionsFactory.as(BigQueryOptions.class);
    System.out.println(options.getClass());
    System.out.println(options.getClass().isInterface());
    System.out.println(PipelineOptionsReflectionSetter.getPipelineOptionsInterface(options));

There will be no exception and the output will be:
    class com.sun.proxy.$Proxy25
    false
    interface org.apache.beam.sdk.io.gcp.bigquery.BigQueryOptions

The reason that I need this method is when the user tries to use reset command, I should set it to its default value. In order to do that, first I try to see what type of pipelineOptions this object is implementing and then using PipelineOptionsFactory I construct an instance of that options and get its default value. 

Currently with our use cases and the expected behavior of PipelineOptionsFactory, this should not fail in any use case. Because the method .as() in PipelineOptions and PipelineOptionsFactory returns a proxy object that implements one interface. I think currently, if the user needs its own piplineOptions they need to declare an interface extending PipelineOptions and then use PipelineOptionsFactory to create an instance of it. 

Hope this makes sense.